src/ifcwrap-csharp/IfcParseWrapper.i

```diff
@@ -0,0 +1,918 @@
+	// Use to correlate to entity_instance.file_pointer, so that we
+	// can trace file ownership of instances on the python side.
+	size_t file_pointer() const {
+		return reinterpret_cast<size_t>($self);
+	}
+
+	IfcUtil::IfcBaseClass* by_guid(const std::string& guid) {
+		return $self->instance_by_guid(guid);
+	}
+	
+	aggregate_of_instance::ptr get_inverse(IfcUtil::IfcBaseClass* e) {
+		return $self->getInverse(e->data().id(), 0, -1);
+	}
+
+	std::vector<int> get_inverse_indices(IfcUtil::IfcBaseClass* e) {
+		return $self->get_inverse_indices(e->data().id());
+	}
+
+	int get_total_inverses(IfcUtil::IfcBaseClass* e) {
+		return $self->getTotalInverses(e->data().id());
+	}
+
+	void write(const std::string& fn) {
+		std::ofstream f(IfcUtil::path::from_utf8(fn).c_str());
+		f << (*$self);
+	}
+
+	std::string to_string() {
+		std::stringstream s;
+		s << (*$self);
+		return s.str();
+	}
+
+	std::vector<unsigned> entity_names() const {
+		std::vector<unsigned> keys;
+		keys.reserve(std::distance($self->begin(), $self->end()));
+		for (IfcParse::IfcFile::entity_by_id_t::const_iterator it = $self->begin(); it != $self->end(); ++ it) {
+			keys.push_back(it->first);
+		}
+		return keys;
+	}
+
+	std::vector<std::string> types() const {
+		const size_t n = std::distance($self->types_begin(), $self->types_end());
+		std::vector<std::string> ts;
+		ts.reserve(n);
+		std::transform($self->types_begin(), $self->types_end(), std::back_inserter(ts), helper_fn_declaration_get_name);
+		return ts;
+	}
+
+	std::vector<std::string> types_with_super() const {
+		const size_t n = std::distance($self->types_incl_super_begin(), $self->types_incl_super_end());
+		std::vector<std::string> ts;
+		ts.reserve(n);
+		std::transform($self->types_incl_super_begin(), $self->types_incl_super_end(), std::back_inserter(ts), helper_fn_declaration_get_name);
+		return ts;
+	}
+
+	std::string schema_name() const {
+		if ($self->schema() == 0) return "";
+		return $self->schema()->name();
+	}
+
+	// %pythoncode %{
+    //     # Hide the getters with read-only property implementations
+    //     header = property(header)
+    //     schema = property(schema_name)
+	// %}
+}
+
+%extend IfcUtil::IfcBaseClass {
+
+	int get_attribute_category(const std::string& name) const {
+		if (!$self->declaration().as_entity()) {
+			return name == "wrappedValue";
+		}
+		
+		{
+		const std::vector<const IfcParse::attribute*> attrs = $self->declaration().as_entity()->all_attributes();
+		std::vector<const IfcParse::attribute*>::const_iterator it = attrs.begin();
+		for (; it != attrs.end(); ++it) {
+			if ((*it)->name() == name) {
+				return 1;
+			}
+		}
+		}
+
+		{
+		const std::vector<const IfcParse::inverse_attribute*> attrs = $self->declaration().as_entity()->all_inverse_attributes();
+		std::vector<const IfcParse::inverse_attribute*>::const_iterator it = attrs.begin();
+		for (; it != attrs.end(); ++it) {
+			if ((*it)->name() == name) {
+				return 2;
+			}
+		}
+		}
+
+		return 0;
```

Comment:
This formatting is cursed. Please fix it

Reply:
i think the formatting is only cursed in githubs PR UI, it looks fine in VSCode